frontend/src/pushapi.js

```diff
@@ -21,49 +17,16 @@ angular.module('clouway-push', [])
     };
 
     /**
-     * Set a method to call for opening of connection.
-     * The method must return a promise.
+     * Sets the backend url connection.
      *
-     * @param {function(subscriber)} method method that will be called for opening of connection.
+     * @param url the new url
      * @returns {*} the provider instance for chaining purposes.
      */
-    this.openConnectionMethod = function (method) {
-      this.connectionMethods.connect = method;
+    this.backendServiceUrl = function(url) {
```

Comment:
Several reasons: 
 * I tried to add new functionality, and it took me more then a hour to figure out what is happening (too much callback hell), which is the force that pushes the tests in some direction
 * Tests are testing mostly the method behaviour, not the overall push api behaviour
 *  Apps that are hooking these methods in the pushApi library, need to know which method when, how, and how many times is called. i.e too much knowledge for the blackbox
 * Test clearance  
For example in the mock version of tests, scope and diferred ($q) need to be used: 
```javascript
 it('call keepAlive after time interval', function () {
      pushApi.openConnection(subscriber);
      connectDeferred.resolve('token');
      rootScope.$digest();

      expect(keepAliveMethod).not.toHaveBeenCalled();
      $interval.flush(keepAliveInterval * 1000);
      expect(keepAliveMethod).toHaveBeenCalledWith(subscriber);
 });
// vs
it('keeps connection alive at specified time interval', function () {
      httpBackend.expectGET('/pushService?subscriber=test-subscriber').respond(200, "token2");

      pushApi.openConnection("test-subscriber");
      httpBackend.flush();

      httpBackend.expectPOST('/pushService?subscriber=test-subscriber').respond(200, "");
      $interval.flush(keepAliveInterval * 1000);

      httpBackend.flush();
});
```
The second version if using the httpBackend and urls stubbing is hardcoded in tests, but all values are passed to production code (event the backendUrl). This could be improved with some helper function like:  initSessionCall("test-subscriber"); bindCall(dummySubscriber, anyCorrelactionId); unbindCall(dummySubscriber, anyCorrelationId) and so on. 

* Error handling is not touched so much and misses things like: lost subscriptions,  keep alive fails, keep alive periodically fails?, unbind errors, bind errors and etc. The caller (app that uses clouwaypush) is not informed about any of these things.